src/main/java/com/google/idea/perf/Autocomplete.kt

```diff
@@ -16,58 +16,101 @@
 
 package com.google.idea.perf
 
+import com.intellij.codeInsight.completion.CompletionParameters
+import com.intellij.codeInsight.completion.CompletionResultSet
+import com.intellij.codeInsight.completion.CompletionUtilCore
+import com.intellij.codeInsight.completion.PrefixMatcher
+import com.intellij.codeInsight.lookup.CharFilter
 import com.intellij.openapi.progress.ProgressManager
+import com.intellij.util.textCompletion.DefaultTextCompletionValueDescriptor
+import com.intellij.util.textCompletion.TextCompletionProvider
 
-data class Suggestion(val name: String, val detail: String?)
+private class AutocompleteMatcher(val pattern: String): PrefixMatcher(pattern) {
+    override fun prefixMatches(name: String): Boolean {
+        return fuzzyMatch(name, pattern) != null
+    }
+
+    override fun cloneWithPrefix(prefix: String): PrefixMatcher {
+        return AutocompleteMatcher(pattern)
+    }
+}
 
-class Autocomplete {
-    private var classNames: List<String> = emptyList()
+class AutocompleteCompletionProvider: TextCompletionProvider {
+    private var classNames = emptyList<String>()
 
+    @Synchronized
     fun setClasses(classes: Collection<Class<*>>) {
         classNames = classes.mapNotNull { it.canonicalName }
```

Comment:
What if someone calls `setClasses()` while another thread is running `predict()`? The `@Synchronized` annotation doesn't help in that case, because the thread running `predict()` does not enter any `@Synchronized` blocks, so it does not [`synchronize-with`](https://docs.oracle.com/javase/specs/jls/se7/html/jls-17.html#jls-17.4.4) the thread running `setClasses()`.

On the other hand, I think adding `@Volatile` to the `classNames` field would work in this case?